### ISSUE TITLE
Generated SETOF table_type functions should reuse the TableRecord type of the referenced table

### ISSUE THREAD

**Issue by ciaccia** [jOOQ/jOOQ#5405]
When I generate the stubs for a PostgreSQL function that returns rows of an existing table, the generator creates a new type. For example, consider the following table and function:

```
CREATE TABLE mytable
(
    id bigserial NOT NULL PRIMARY KEY,
    isactive boolean,
    val1 int,
    val2 int,
    val3 int,
    val4 int,
    val5 int,
    val6 int,
    val7 int,
    val8 int,
    prop1 text,
    prop2 text,
    prop3 text,
    prop4 text,
    prop5 text,
    prop6 text,
    prop7 text,
    prop8 text,
    -- Default entity fields
    created_by text NOT NULL DEFAULT 'db_script',
    modified_by text NOT NULL DEFAULT 'db_script',
    created_at timestamp with time zone DEFAULT now(),
    modified_at timestamp with time zone DEFAULT now(),
    versionnr bigint NOT NULL    
);

CREATE OR REPLACE FUNCTION testfunction() 
RETURNS SETOF mytable AS $$
    select t.* from mytable t where t.isactive = true;
$$ LANGUAGE SQL;
```

The stub generated for testfunction uses a record named `TestfunctionRecord` which is not the same as `MytableRecord` and completely _"incompatibile"_ despite having exaclty the same fields. Ideally testfunction should use `MytableRecord` records as well.

**Comment by lukaseder** [comment]
Thank you very much for your suggestion. Support for such `SETOF` functions was added in jOOQ 3.6 with https://github.com/jOOQ/jOOQ/issues/3376. I'm not sure anymore why the decision was made to generate separate `Record` types for these. The table type information is present on:

```
information_schema.routines.type_udt_catalog
information_schema.routines.type_udt_schema
information_schema.routines.type_udt_name
```

This information is also used to join the `columns` table to get a hold of the actual columns returned.

At first sight, it certainly makes sense to proceed as you suggest. This might even be useful to store back records into the database. Will update as soon as I know more about the historic context.

**Comment by lukaseder** [comment]
Looking into this now as I've just implemented a related feature request:
https://github.com/jOOQ/jOOQ/issues/7503

**Comment by lukaseder** [comment]
> Will update as soon as I know more about the historic context.

I don't think there was anything specific in terms of historic context. It was just an overlooked opportunity.

While it would be possible to make this configurable, I tend to think that a breaking change is better:

- It's not the most popular feature within jOOQ or within PostgreSQL
- When users use `RETURNS SETOF some_type`, then the expectation is to have that precise type in the generated code, not what jOOQ is doing now. What jOOQ is doing now is fine for the `RETURNS TABLE (...)` syntax, where a new table type is generated for a single function specification.
- We're already doing what this issue is requesting for Oracle, when people use `%ROWTYPE` declarations, which are equivalent to this PostgreSQL feature. I've received positive feedback about the Oracle integration's behaviour.

**Comment by lukaseder** [comment]
A related change will be when `SETOF` types reference UDT types: https://github.com/jOOQ/jOOQ/issues/14298

**Comment by lukaseder** [comment]
Fixed for jOOQ 3.18.0

**Comment by lukaseder** [comment]
This change produced a subtle regression when:

- The returned table has client side computed columns
- `<recordsImplementingRecordN/>` is active

See:

- https://github.com/jOOQ/jOOQ/issues/14830
